RootEventOutputer.cc

```diff
@@ -0,0 +1,308 @@
+  std::cout << "  end of job file close time: "<<closeTime.count()<<"us\n";
+                                                                                         
+  summarize_serializers(serializers_);
+}
+
+
+
+void RootEventOutputer::output(EventIdentifier const& iEventID, SerializeStrategy const& iSerializers, std::vector<uint32_t>const& iBuffer) {
+  if(firstTime_) {
+    writeMetaData(iSerializers);
+    firstTime_ = false;
+  }
+  //using namespace std::string_literals;
+  
+  //std::cout <<"   run:"s+std::to_string(iEventID.run)+" lumi:"s+std::to_string(iEventID.lumi)+" event:"s+std::to_string(iEventID.event)+"\n"<<std::flush;
+  
+  //writeEventID(iEventID);
+  eventID_ = iEventID;
+  eventBlob_ = std::move(iBuffer);
+  //std::cout <<"Event "<<eventID_.run<<" "<<eventID_.lumi<<" "<<eventID_.event<<std::endl;
+  //std::cout <<"buffer size "<<eventBlob_.size();
+  //for(auto b: eventBlob_) {
+  //  std::cout <<"   "<<b<<std::endl;
+  //}
+  eventsTree_->Fill();
+  /*
+    for(auto& s: iSerializers) {
+    std::cout<<"   "s+s.name()+" size "+std::to_string(s.blob().size())+"\n" <<std::flush;
+    }
+  */
+}
+
+void RootEventOutputer::writeMetaData(SerializeStrategy const& iSerializers) {
+
+  std::vector<std::pair<std::string, std::string>> typeAndNames;
+  for(auto const& s: iSerializers) {
+    std::string type(s.className());
+    std::string name(s.name());
+    typeAndNames.emplace_back(std::move(type), std::move(name));
+  }
+
+  int objectSerializationUsed = static_cast<std::underlying_type_t<Serialization>>(serialization_);
+
+
+  std::string compression;
+  {
+    //Compression type used
+    // note want exactly 4 bytes so sometimes skip trailing \0
+    switch(compression_) {
+    case Compression::kNone:
+      {
+        compression = "None";
+	break;
+      }
+    case Compression::kLZ4:
+      {
+        compression = "LZ4";
+	break;
+      }
+    case Compression::kZSTD:
+      {
+        compression = "ZSTD";
+	break;
+      }
+    }
+  }
+
+  TTree* meta = new TTree("Meta", "File meta data", 0, &file_);
+  meta->Branch("DataProducts",&typeAndNames, 0, 0);
+  meta->Branch("objectSerializationUsed",&objectSerializationUsed);
+  meta->Branch("compressionAlgorithm",&compression,0,0);
+
+  meta->Fill();
+
+}
+
+std::vector<uint32_t> RootEventOutputer::writeDataProductsToOutputBuffer(SerializeStrategy const& iSerializers) const{
+  //Calculate buffer size needed
+  uint32_t bufferSize = 0;
+  for(auto const& s: iSerializers) {
+    bufferSize +=1+1;
+    auto const blobSize = s.blob().size();
+    bufferSize += bytesToWords(blobSize); //handles padding
+  }
+  //initialize with 0
+  std::vector<uint32_t> buffer(size_t(bufferSize), 0);
+  
+  {
+    uint32_t bufferIndex = 0;
+    uint32_t dataProductIndex = 0;
+    for(auto const& s: iSerializers) {
+      //std::cout <<"  write: "<<s.name()<<std::endl;
+      buffer[bufferIndex++]=dataProductIndex++;
```

Comment:
In a future iteration I think I'll

- drop storing of the product index
- remove storing the size with the product and instead at the beginning of the buffer create a table of offsets saying where each data product begins. The table would be 1 larger than the number of products and have the buffer length as the last entry. The size of a data product is then calculated by doing a subtraction of neighboring entries in the table.